### PR TITLE
Update job priority only for processing and production task types

### LINKED ISSUE #10445: Adopt a different method for updating jobs in JobUpdater
**Impact of the new feature**
WMAgent

**Is your feature request related to a problem? Please describe.**
Depending on the status of the schedd and amount of jobs in the system, running condor_qedit becomes a heavy action and it should be avoided if possible. Right now, when we need to update the priority for a workflow, we actually run condor_qedit matching a given WMAgent_SubTaskName, as far as I remember.

**Describe the solution you'd like**
Instead of running many condor_qedit for the very same workflow - passing a different task path every time - I think we could actually run a single call for each workflow, with the following format:
```
condor_qedit -const 'JobStatus=?=1 && (CMS_JobType =?= "Production" || CMS_JobType =?= "Processing")' JobPrio JOB_PRIO_VALUE
```

**Describe alternatives you've considered**
Not a critical development, so we could not make any changes as well.

**Additional context**
As one can see in this component log, a whole cycle updating 8 workflows too around 8 minutes, fairly slow in my opinion:
```
2021-04-13 18:23:48,012:140261058021120:INFO:JobUpdaterPoller:Found 8 workflows to update in workqueue
2021-04-13 18:24:44,926:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_TSG-Run3Winter21GS-00006__v1_T_210403_135412_3515/TSG-Run3Winter21GS-00006_0/TSG-Run3Winter21DRMiniAOD-00002_1MergeMINIAODSIMoutput
2021-04-13 18:24:44,926:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:27:10,536:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_TOP-RunIISummer20UL18wmLHEGEN-00054__v1_T_201206_222616_3749/TOP-RunIISummer20UL18wmLHEGEN-00054_0/TOP-RunIISummer20UL18RECO-00052_0MergeAODSIMoutput
2021-04-13 18:27:10,536:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:28:14,551:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_TOP-RunIISummer20UL18wmLHEGEN-00061__v1_T_201206_222729_9983/TOP-RunIISummer20UL18wmLHEGEN-00061_0/TOP-RunIISummer20UL18RECO-00059_0MergeAODSIMoutput
2021-04-13 18:28:14,551:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:29:06,090:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_TOP-RunIISummer20UL18wmLHEGEN-00061__v1_T_201206_222729_9983/TOP-RunIISummer20UL18wmLHEGEN-00061_0/TOP-RunIISummer20UL18MiniAOD-00059_0MergeMINIAODSIMoutput
2021-04-13 18:29:06,091:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:30:57,409:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_BTV-RunIISummer20UL16GENAPV-00028__v1_T_210127_101322_3364/BTV-RunIISummer20UL16GENAPV-00028_0/BTV-RunIISummer20UL16RECOAPV-00015_0MergeAODSIMoutput
2021-04-13 18:30:57,409:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:31:47,324:140261058021120:ERROR:SimpleCondorPlugin:Failed to update JobPrio for WMAgent_SubTaskName=/cmsunified_task_BTV-RunIISummer20UL16GENAPV-00028__v1_T_210127_101322_3364/BTV-RunIISummer20UL16GENAPV-00028_0/BTV-RunIISummer20UL16MiniAODAPV-00015_0MergeMINIAODSIMoutput
2021-04-13 18:31:47,324:140261058021120:ERROR:SimpleCondorPlugin:Unable to edit jobs matching constraint
Traceback (most recent call last):
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/cms/wmagent/1.4.5.patch1/lib/python2.7/site-packages/WMCore/BossAir/Plugins/SimpleCondorPlugin.py", line 485, in updateJobInformation
    schedd.edit(constraint, 'JobPrio', classad.Literal(newPriority))
  File "/data/srv/wmagent/v1.4.5.patch1/sw/slc7_amd64_gcc630/external/condor/8.9.7-comp/lib/python2.7/site-packages/htcondor/_lock.py", line 75, in wrapper
    rv = func(*args, **kwargs)
RuntimeError: Unable to edit jobs matching constraint
2021-04-13 18:31:56,265:140261058021120:INFO:JobUpdaterPoller:Updating 61 workflows in WMBS.
2021-04-13 18:31:56,286:140261058021120:INFO:JobUpdaterPoller:Priorities were synchronized, wait until the next cycle
2021-04-13 18:31:56,286:140261058021120:INFO:BaseWorkerThread:JobUpdaterPoller took 491.051 secs to execute
```

### PR DESCRIPTION
**Pull request by khurtado**
Fixes #10445

#### Status
 not-tested

#### Description
Update job priority only for processing/production types

#### Is it backward compatible (if not, which system it affects?)
YES

### CODE REVIEW

**Comment by amaltaro** [comment]
@khurtado Kenyi, I know you are still working on it, but I just wanted to say that this will still execute a condor_qedit for every single Processing/Production task in the workflow (for StepChains it would be a single time, but for TaskChains it could be many). 

My suggestion would be to move away of the `WMAgent_SubTaskName` constraint, and instead execute this job priority edit only for a tuple of RequestName and JobType.

**Comment by khurtado** [comment]
@amaltaro Okay, I modified it in a way it would only invoke the Bossair updateJobInformation method once per request,  but kept the (request, task) tuple for the other methods that are invoked in e.g.: WQ

File: src/python/WMComponent/JobUpdater/JobUpdaterPoller.py [hunk 1]
```diff
@@ -157,9 +158,11 @@ def synchronizeJobPriority(self):
                 self.workqueue.updatePriority(workflow, requestPriority)
                 # Check if there are executing jobs for this particular task
                 if self.executingJobsDAO.execute(workflow, workflowEntry['task']) > 0:
-                    self.bossAir.updateJobInformation(workflow, workflowEntry['task'],
-                                                      requestPriority=priorityCache[workflow],
-                                                      taskPriority=workflowEntry['task_priority'])
+                    # Only update once per request workflow
+                    if workflow not in updatedRequest:
+                        updatedRequest += [workflow]
```
Comment (amaltaro) [inline src/python/WMComponent/JobUpdater/JobUpdaterPoller.py:163]:
```suggestion
                        updatedRequest.append(workflow)
```
how about this?

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 2]
```diff
@@ -475,18 +475,21 @@ def updateJobInformation(self, workflow, task, **kwargs):
...
         schedd = htcondor.Schedd()
 
-        if 'taskPriority' in kwargs and 'requestPriority' in kwargs:
-            newPriority = int(kwargs['requestPriority']) + int(kwargs['taskPriority'] * self.maxTaskPriority)
+        if 'requestPriority' in kwargs:
+            newPriority = int(kwargs['requestPriority'])
             try:
-                constraint = "WMAgent_SubTaskName =?= %s" % classad.quote(str(task))
-                constraint += " && WMAgent_RequestName =?= %s" % classad.quote(str(workflow))
+                constraint = " && WMAgent_RequestName =?= %s" % classad.quote(str(workflow))
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py:487]:
```suggestion
                constraint = "WMAgent_RequestName =?= %s" % classad.quote(str(workflow))
```

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py [hunk 3]
```diff
@@ -475,18 +475,21 @@ def updateJobInformation(self, workflow, task, **kwargs):
...
                 constraint += " && JobPrio =!= %d" % newPriority
+                constraint += " && stringListMember(CMS_JobType, %s) " % classad.quote(str("Production, Processing"))
```
Comment (amaltaro) [inline src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py:489]: Kenyi, did you manage to test this line?

**Review by amaltaro — CHANGES_REQUESTED** [review]
Thanks for making these changes, Kenyi. Please find some comments inline.

On a different approach, I considered changing the DAO `ListForJobUpdater` such that it returns a single record by workflow, since it's only used by this component. But then it would require further changes to the DAO listing executing jobs for a given workflow+task.

File: src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py (hunk 3 above)
Comment (khurtado) [reply src/python/WMCore/BossAir/Plugins/SimpleCondorPlugin.py:489]: Yes, I tested that in an isolated environment

**Comment by amaltaro** [comment]
test this please

**Review by amaltaro — APPROVED** [review]

**Comment by amaltaro** [comment]
Thanks, Kenyi! I performed a few tests with this patch and it seems to be working fine.
I did notice though an - unrelated - issue with this line:
```
self.workqueue.updatePriority(workflow, requestPriority)
```
which gets executed X times for the very same workflow, where X is the number of tasks. I can follow this up with another PR though.

@khurtado could you please squash those commits into a single one? Thanks

**Comment by khurtado** [comment]
@amaltaro Done!

**Comment by amaltaro** [comment]
test this please

**Comment by amaltaro** [comment]
Thanks, Kenyi!

**Pull request merged**